src/backend/distributed/planner/multi_colocated_subquery_planner.c

```diff
@@ -0,0 +1,783 @@
+/*
+ * multi_colocated_subquery_planner.c
```

Comment:
I'd be inclined to go for shard_pruning.c or such - that's what I'd added in my faster shard pruning stuff, and this pretty much related.